File: server/app/services/geocoding.py
import httpx


class GeocodingError(Exception):
    pass


async def geocode_zip(zip_code: str) -> dict:
    """Resolve a US zip code to lat/lon using Zippopotam.us."""
    url = f"https://api.zippopotam.us/us/{zip_code}"
    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(url, timeout=10)
    except httpx.HTTPError as e:
        raise GeocodingError(f"Geocoding service unavailable: {e}")

    if response.status_code == 404:
        raise GeocodingError(f"No location found for zip code: {zip_code}")
    if response.status_code != 200:
        raise GeocodingError(f"Geocoding failed with status {response.status_code}")

    data = response.json()
    place = data["places"][0]
    lat = float(place["latitude"])
    lon = float(place["longitude"])
    location_name = f"{place['place name']}, {place['state abbreviation']}"
    return {"lat": lat, "lon": lon, "location_name": location_name}

File: src/server/services/geocoding.ts
export class GeocodingError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'GeocodingError';
  }
}

export async function geocodeZip(
  zipCode: string,
): Promise<{ lat: number; lon: number; locationName: string }> {
  const url = `https://api.zippopotam.us/us/${zipCode}`;

  let response: Response;
  try {
    response = await fetch(url, { signal: AbortSignal.timeout(10000) });
  } catch (e) {
    throw new GeocodingError(
      `Geocoding service unavailable: ${e instanceof Error ? e.message : String(e)}`,
    );
  }

  if (response.status === 404) {
    throw new GeocodingError(`No location found for zip code: ${zipCode}`);
  }
  if (!response.ok) {
    throw new GeocodingError(
      `Geocoding failed with status ${response.status}`,
    );
  }

  const data = await response.json();
  const place = data.places[0];
  const lat = parseFloat(place.latitude);
  const lon = parseFloat(place.longitude);
  const locationName = `${place['place name']}, ${place['state abbreviation']}`;

  return { lat, lon, locationName };
}
